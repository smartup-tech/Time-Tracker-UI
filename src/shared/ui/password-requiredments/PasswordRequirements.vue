<script lang="ts" setup>
import { computed } from 'vue';

import {
  hasLowerCaseLetters,
  hasNumbers,
  hasUpperCaseLetters,
} from '@/shared/lib';
import { LAIcon } from '@/shared/ui';

type Props = {
  password?: string;
};

type Rule = {
  isValid: boolean;
  text: string;
};

const MIN_LENGTH = 5;
const MAX_LENGTH = 30;

const props = defineProps<Props>();

const lengthOk = computed<boolean>(() => {
  const length = props.password ? String(props.password).length : 0;
  return length >= MIN_LENGTH && length <= MAX_LENGTH;
});

const spacesOk = computed<boolean>(() =>
  props.password ? !/\s/.test(props?.password) : true
);

const numbersOk = computed<boolean>(() =>
  props.password ? hasNumbers(props.password) : true
);

const upperCaseOk = computed<boolean>(() =>
  props.password ? hasUpperCaseLetters(props.password) : true
);

const lowerCaseOk = computed<boolean>(() =>
  props.password ? hasLowerCaseLetters(props.password, 2) : true
);

const withFeedback = computed<boolean>(() => Boolean(props.password));

const getIcon = (isOk: boolean): string =>
  props.password ? (isOk ? 'check-circle' : 'times-circle') : 'circle';

const rules = computed<Rule[]>(() => [
  {
    isValid: lengthOk.value,
    text: `Длина от ${MIN_LENGTH} до ${MAX_LENGTH} символов`,
  },
  {
    isValid: upperCaseOk.value,
    text: 'Минимум одна латинская заглавная буква',
  },
  {
    isValid: lowerCaseOk.value,
    text: 'Минимум две латинских строчных буквы',
  },
  {
    isValid: numbersOk.value,
    text: 'Минимум одна цифра',
  },
  {
    isValid: spacesOk.value,
    text: 'Без пробелов',
  },
]);
</script>

<template>
  <div
    :class="[
      'password-requirements',
      withFeedback && 'password-requirements_with-feedback',
    ]"
  >
    Требования к паролю:
    <ul class="password-requirements__list">
      <li
        v-for="(rule, index) in rules"
        :key="index"
        class="password-requirements__list-item"
      >
        <LAIcon
          v-if="withFeedback"
          :class="[
            'password-requirements__icon',
            rule.isValid
              ? 'password-requirements__icon_success'
              : 'password-requirements__icon_error',
          ]"
          :icon="getIcon(rule.isValid)"
          size="small"
        />
        {{ rule.text }}
      </li>
    </ul>
  </div>
</template>
<style lang="scss">
.password-requirements {
  &__list {
    margin: 8px 0 0;
    padding: 0;
  }

  &__list-item {
    position: relative;
    margin: 0 0 4px 20px;

    .password-requirements_with-feedback & {
      list-style: none;
    }
  }

  &__icon {
    position: absolute;
    left: -22px;
    top: 2px;

    &_success {
      color: var(--color-success);
    }

    &_error {
      color: var(--color-error);
    }
  }
}
</style>
