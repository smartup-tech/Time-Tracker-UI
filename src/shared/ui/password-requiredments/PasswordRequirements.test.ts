import { mount } from '@vue/test-utils';
import PasswordRequirements from './PasswordRequirements.vue';

describe('Password Requirements', () => {
  test('match snapshot', async () => {
    const wrapper = mount(PasswordRequirements);
    expect(wrapper.element).toMatchSnapshot();
  });
});
