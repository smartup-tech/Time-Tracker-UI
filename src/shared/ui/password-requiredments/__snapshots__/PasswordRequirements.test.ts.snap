// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Password Requirements > match snapshot 1`] = `
<div
  class="password-requirements"
>
   Требования к паролю: 
  <ul
    class="password-requirements__list"
  >
    
    <li
      class="password-requirements__list-item"
    >
      <!--v-if-->
       Длина от 5 до 30 символов
    </li>
    <li
      class="password-requirements__list-item"
    >
      <!--v-if-->
       Минимум одна латинская заглавная буква
    </li>
    <li
      class="password-requirements__list-item"
    >
      <!--v-if-->
       Минимум две латинских строчных буквы
    </li>
    <li
      class="password-requirements__list-item"
    >
      <!--v-if-->
       Минимум одна цифра
    </li>
    <li
      class="password-requirements__list-item"
    >
      <!--v-if-->
       Без пробелов
    </li>
    
  </ul>
</div>
`;
